ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Adversarial.hs

```diff
@@ -0,0 +1,521 @@
+    -- * Testing
+  , AdversarialViolation (BadAnchor, BadCount, BadRace)
+  , AnchorViolation (HonestActiveMustAnchorAdversarial)
+  , ChainSchema (ChainSchema)
+  , RaceViolation (AdversaryWonRace, rvAdv, rvHon)
+  , checkAdversarialChain
+  ) where
+
+import           Control.Applicative ((<|>))
+import           Control.Monad (void, when)
+import qualified Control.Monad.Except as Exn
+import           Data.Proxy (Proxy (Proxy))
+import qualified System.Random.Stateful as R
+import qualified Test.Ouroboros.Consensus.ChainGenerator.BitVector as BV
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Counting as C
+import           Test.Ouroboros.Consensus.ChainGenerator.Honest
+                     (ChainSchema (ChainSchema))
+import           Test.Ouroboros.Consensus.ChainGenerator.Params (Asc,
+                     Delta (Delta), Kcp (Kcp), Scg (Scg))
+import qualified Test.Ouroboros.Consensus.ChainGenerator.RaceIterator as RI
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Slot as S
+import           Test.Ouroboros.Consensus.ChainGenerator.Slot
+                     (E (ActiveSlotE, EmptySlotE, SlotE))
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Some as Some
+
+-----
+
+data AnchorViolation =
+    -- | An honest active slot must immediately precede the adversarial interval
+    HonestActiveMustAnchorAdversarial
+  |
+    -- | The were not exactly 'arPrefix' many active slots preceding @adv@
+    WrongNumberOfHonestPredecessors
+  deriving (Eq, Read, Show)
+
+-- | A violation of the Extended Praos Race Assumption
+--
+-- INVARIANT: @'C.windowLast' 'rvAdv' < 'C.windowLast' 'rvHon' + 'Delta'@
+--
+-- INVARIANT: @'C.windowStart' 'rvHon' <= 'C.windowStart' 'rvAdv'@
+data RaceViolation hon adv = AdversaryWonRace {
+    -- | The adversarial race window
+    rvAdv :: !(RI.Race adv)
+  ,
+    -- | The honest race window
+    rvHon :: !(RI.Race hon)
+  }
+  deriving (Eq, Read, Show)
+
+data AdversarialViolation hon adv =
+    BadAnchor !AnchorViolation
+  |
+    BadCount
+  |
+    BadRace !(RaceViolation hon adv)
+  deriving (Eq, Read, Show)
+
+-- | Check the chain matches the given 'AdversarialRecipe', especially the Extended Praos Race Assumption
+--
+-- Definition of a /Praos Race Window/. Such a window includes exactly 'Delta'
+-- slots after the @k+1@st honest block in the window.
+--
+-- Definition of the /Praos Race Assumption/. We assume every adversarial chain
+-- contains at most @k@ blocks in the Praos Race Window anchored at the
+-- intersection.
+--
+-- Definition of the /Extended Praos Race Assumption/. Beyond the Praos Race
+-- Assumption, we assume that every adversarial chain also has at most @k@
+-- blocks in every subsequent Praos Race Window that ends within 'Delta' of the
+-- Stability Window anchored at the first adverarial block. (Crucially, these
+-- are the Praos Race Windows in which the the adversary could not yet have
+-- drastically accelerated its rate of election.)
```

Comment:
And this is the property that I think we should replace with "Length of competing chains". The two properties are similar, but LoCC does not care about the stability window (that is the sole concern of Chain Growth instead).

Reply:
:+1: on the general impetus to use the terminology from the test plan also here in the code, and indeed they are similar.

A difference is that the LoCC is concerned just with the first race right after an intersection, while the EPRA is concerned with additional races beyond that (that is also why the EPRA has to talk about the stability window in contrast to the LoCC, as that constraints the adversary in "speeding up"). As you say, this is to rule out "improbable" leader schedules. We could either incorporate that into the LoCC in the test plan, or have an additional property that captures the guarantees that the EPRA currently gives on top of LoCC+DoCC.

Reply:
Ah, I think I'm just understanding EPRA enough to recognize that it is accomplishing the same as the strengthened LoCC.

Reply:
In c06d54a I renamed EPRA to Adversarial chains in stability windows.

I'm changing the name to make it more precise, and also because it doesn't obviously imply the Praos Race Assumption. In particular the implication is unclear when the Praos Race Window is larger than the stability window.